Allow upcasting and downcasting of numerical values in SQL predicates.

amrishlal opened the pull request:
**Problem**
This PR is motivated by a discussion thread in Apache Pinot # general slack channel (see Dec 18th, 2020 post by Will Briggs), where we noticed that the query `SELECT COUNT(*) FROM myTable WHERE eventTimestamp <= NOW() - 10000` was not working. Upon investigation, it was discovered that:

- This was an issue with upcasting and downcasting of numerical values in predicate.
- Any query that compared two different numerical type wouldn't work due the type casting issue.
Example: `SELECT count(*) FROM testTable WHERE intColumn > 15.1`.
- Constant expressions are evaluated by Broker and converted into decimal value string, so a query such as:
    `SELECT count(*) FROM testTable WHERE intColumn = 5 - 4`
is transformed into the query:
    `SELECT count(*) FROM testTable WHERE intColumn = 1.0` on the server side.
Since the query compares two different numerical types, it fails to evaluate.

**Solution**
The basic approach taken here is to rewrite the predicate at compile time to an equivalent predicate that will work against column data type (see rewrite() method in Predicate interface and its implementation). Rewriting the predicate at compile time is preferable as it causes no performance impact during runtime. Since we are only dealing with numerical values here, this PR touched `EQ`, `NOT_EQ`, `IN`, `NOT_INT`, and `RANGE` predicates.

All numerical string values used in predicates are first parsed as java.lang.BigDecimal and then downcasted to numerical type of the column (See changes in `ColumnValueSegmentPruner`). The reason we use BigDecimal, but not Double is because long to double to long conversation is lossy (i.e does not give back the original long value).

Changes in functionality and expected results were confirmed by running queries against MySQL and PostgreSQL databases.

**Example 1** (RANGE)
`SELECT count(*) FROM testTable where intColumn > 3_000_000_000`
We check MIN/MAX integer bounds in the RangePredicate and rewrite the predicate at compile time into an equivalent form that would work against integer column even though the literal value is of long data type. See `RangePredicate.rewrite` method for more detail.

**Example 2** (EQ, NOT_EQ)
The query:
  `SELECT count(*) FROM testTable WHERE intColumn = 45.1`
will always evaluate to FALSE since no integer column can have decimal value 45.1. Similarly the query:
  `SELECT count(*) FROM testTable WHERE intColumn != 45.1`
will always evaluate to TRUE.

We can tell whether such predicates will evaluate to true or false at compile time, so there is no need to evaluate such predicates during runtime (See `EqPredicate.rewrite`, `NotEqPredicate.rewrite`). A boolean `_precomputed` member variable was added in BasePredicate to keep track of such predicates. Precomputation, where possible, makes predicate evaluation very fast.

**Example 3** (IN, NOT_IN)
`select count(*) from scores where intColumn IN (12, 1.0, 25.01, 40.0, -12.0)`
`select count(*) from scores where intColumn NOT IN (12, 1.0, 25.01, 40.0, -12.0)`
This query gets rewritten to:
`select count(*) from scores where intColumn IN (12, 1, 40, -12)`
`select count(*) from scores where intColumn NOT IN (12, 1, 40, -12)`

Note that `24.01` was removed from IN predicate because an integer column can never have a decimal value. Also the value `1.0` was converted to `1` and then value `-12.0` was converted to `-12` since these decimal values are compatible with integer values.

**Example 4** (HAVING CLAUSE)
Upcasting and downcasting of numerical predicates is also supported in HAVING clause. See examples in `HavingClauseNumericalPredicateTest.java`.


**Possible Future Work**
Predicate parsing and compilation can be made more efficient. However, I feel that this is outside the scope of this PR (and would make this PR extremely large if we were to do it here).

There may be other predicates that can be evaluated at compile time. We may be able to utilize predicate pre-computation further to speed up runtime evaluation of predicates.

There may be existing query correctness issues related to 1) long to double conversion and 2) how certain numerical calculations handle overflow and underflow.

amrishlal commented:
@mayankshriv @Jackie-Jiang This is the PR for dealing with timestamp arithmetic evaluation discussed in Apache Pinot slack channel. Please take a look.

amrishlal commented:
Question: Do we need to modify type handling in ConstantValueLongDictionary, ConstantValueIntDictionary, ConstantValueFloatDictionary, and ConstantValueDoubleDictionary (like what was done for IntDictionary, FloatDictionary, and LongDictionary)? What are ConstantValueXXXDictionary classes used for?

mcvsubbu commented:
> Question: Do we need to modify type handling in `ConstantValueLongDictionary`, `ConstantValueIntDictionary`, `ConstantValueFloatDictionary`, and `ConstantValueDoubleDictionary` (like what was done for `IntDictionary`, `FloatDictionary`, and `LongDictionary`)? What are ConstantValueXXXDictionary classes used for?

These are used for addiing a dictionary to a column in a CONSUMING segment, when a column is added as the segment is being consumed. We reflect the default value (coded in the constantValue*Dictionary) for the life of the consuming segment.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/FloatDictionary.java:
```diff
@@ -30,7 +31,24 @@ public FloatDictionary(PinotDataBuffer dataBuffer, int length) {
-    return binarySearch(Float.parseFloat(stringValue));
+    // First convert string to BigDecimal and then downcast to int. This allows type conversion from any compatible
+    // numerical value represented as string to an int value.
+    BigDecimal bigDecimal = new BigDecimal(stringValue);
```
This is a FloatDictionary, so we can safely assume that the caller stack ensure it is called with a stringValue that is expected to be float parsed. I don't think we should convert it into BigDecimal.

amrishlal replied:
This doesn't appear to be the case. For example if the query is:
    `SELECT count(*) from mytable WHERE intColumn < 3000000000.453`
then `stringValue` gets set to 3000000000.453 and the current code would throw exception because 3000000000.453 is not a valid integer.

After the change, the above query will evaluate successfully (which is expected SQL behavior). Same with other queries such as:
    `SELECT count(*) from mytable WHERE intColumn = 5.0

mayankshriv replied:
What is the expected behavior from other SQL engines if the value does not match the column data type: `intColumn < 3000000000.453`?

amrishlal replied:
The expected behavior is that the query will still evaluate and produce results as long as the numerical value can be upcasted/downcasted to the column type. For example, the following works on MySQL and PostgreSQL.

```
create table test(intColumn int, floatColumn float);
insert into test values(5, 5.5);
insert into test values(10, 15.56);

select * from test where intColumn < 300.453;
select * from test where intColumn < 300000000000000000000000.453;
select * from test where floatColumn > 6;
```
300000000000000000000000.453 is outside the bounds of Integer datatype, so it can't be represented as an Integer value, but gets parsed to something that can be compared to an integer.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/IntDictionary.java:
```diff
@@ -30,7 +31,24 @@ public IntDictionary(PinotDataBuffer dataBuffer, int length) {
-    return binarySearch(Integer.parseInt(stringValue));
+    // First convert string to BigDecimal and then downcast to int. This allows type conversion from any compatible
+    // numerical value represented as string to an int value.
+    BigDecimal bigDecimal = new BigDecimal(stringValue);
```
Same here, and all other data type dependent dictionaries.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/query/pruner/ColumnValueSegmentPruner.java:
```diff
@@ -239,17 +240,24 @@ private boolean pruneRangePredicate(IndexSegment segment, RangePredicate rangePr
       switch (dataType) {
         case INT:
-          return Integer.valueOf(stringValue);
+          return (new BigDecimal(stringValue)).intValue();
```
Using BigDecimal may have performance penalty. If we already know the data type for the column, why not down-cast early on?

amrishlal replied:
The problem is that currently `stringValue` could be anything (Integer, Float, Double, Long, Text), so we need: 1) a way to tell if stringValue represents a number (and for this 'stringValue' is parsed as a BigDecimal which parses a wide variety of numerical formats), and 2) if stringValue is a number then we need to upcast/downcast that number to same value as the column type (and hence the downcasting) for predicate evaluation.

mayankshriv replied:
I highly suspect this will have potential impact on performance (especially for high throughput cases). I think the original case that you were trying to address was a case where Pinot returned a specific data type `double` as part of a UDF that was being compared to a column with data type `long`. Also, from what I recall, the question was whether the UDF itself should have returned long.
In the example you mentioned, the RHS is provided by the user, and not really the output of UDF.

My recommendation is to first find out how other SQL engines behave in such cases, as in:
- Do they throw error for incompatible comparison?

amrishlal replied:
I think this has more to do with how we are evaluating predicates on Calcite / Broker side rather than UDF. For example, the following query on the broker side:
  `SELECT * FROM mytable WHERE intColumn > 5 - 4`
will get send to the server as:
  `SELECT * FROM mytable WHERE intColumn >1.0`
which requires comparing an int column with a float value. Same thing carries over to queries such as:
  ` SELECT * FROM mytable WHERE longColumn > NOW() - 1000`
which are sent to the server side as:
  `SELECT * FROM mytable WHERE longColumn > 1.610609874257E12`

> Do they throw error for incompatible comparison?
No, as long as the numerical types being used in predicates can be compared, the query seems to run fine and return results. I tried several different variations on both MySQL and PostgreSQL to find an exception or limitation to this rule, but could not find one.

> Downcast RHS before comparison
> Upcast LHS before comparison
I think as long as RHS and LHS can be compared, casting may not be needed. For example, in our case 300000000000000000.453 is obviously bigger than any integer value so the predicate "intColumn < 300000000000000000.453" will always be true and we don't need to cast to integer, but it still needs to parse as a numerical value to carry out some sort of comparison.

If we have the predicate like `intColumn < 3.453` then we parse "3.453" to a BigInteger value and then cast it into integer for carrying out binary search. This is very similar to what would happen if we had a predicate `intColumn < 3`. Here we would parse / cast "3" as an integer for binary search and if for some reason 3 didn't parse as an integer we would throw an exception. In either case 3.453 or 3 is being parsed as biginteger or integer only once just like any other literal in the query and this parsing is independent of the number of values in the column.

Jackie-Jiang commented:
Per the performance concern, can we always parse the String as double?

amrishlal commented:
> Per the performance concern, can we always parse the String as double?

The main reason why I used BigDecimal was that long to double to long conversion does not give back the original value since both are 64bit. I will look into this a bit more, but it is probably safer to lean in the direction of query correctness as compared to performance.

amrishlal commented:
@mayankshriv @mcvsubbu @siddharthteotia @Jackie-Jiang @fx19880617 This PR is ready for review again. Correctness issues are now fixed and I have added test cases to ensure things are working properly. Please review. For examples see PR description and test cases.

Jackie-Jiang reviewed (changes requested):
Sorry I have to block this PR because I think it is not done correctly.

Currently this rewrite is not done at compile time, but during query execution on every segment (triggered in `PredicateEvaluatorProvider.getPredicateEvaluator()` during the per-segment query planning).
Also, predicates are shared among all segments, so it must be immutable, or we will run into concurrency issue.
The per-value `_precomputed` check could also impact the performance.

This rewrite should be performed on the broker side (within `BaseBrokerRequestHandler` class) where we have access to the schema of the table. You can implement similar to the `QueryOptimizer` class, or make it part of that.

amrishlal commented:
@Jackie-Jiang @siddharthteotia As discussed offline, we will move the existing rewrite logic in this PR from Server side to Broker side. This may not support expressions of type `function(column) <operator> <literal>` since the LHS type is being resolved at runtime, but that is something we will look into more detail after adding support for `<column> <operator> <literal>` type expressions. As a separate item, we will look at the scalar function overloading logic at some point in future (specially with respect to whether templates can be used to generate overloaded scalar functions at compile time).

amrishlal commented:
Closing PR as changes have been incorporated on the Broker side using other PRs.
